### PR TITLE
PSF fitting methods

### LINKED ISSUE #781: PSF fitting for accurate astrometry in Source Detection
_Issue [RCAL-609](https://jira.stsci.edu/browse/RCAL-609) was created on JIRA by [Brett Morris](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bmmorris):_

RCAL-573 investigated PSF fitting for accurate astrometry, now implement the "optimal" solutions according to [the investigation summary from the previous ticket.](https://innerspace.stsci.edu/pages/viewpage.action?spaceKey=SCSB&title=Roman+astrometry+via+PSF+fitting+for+Source+Detection+Step)

Add narrative docs for PSF fitting methods either in Source Detection, or link to them there.

### PR DESCRIPTION
**Pull request by bmorris3**
### Description

This PR introduces methods for fitting model PSFs computed with WebbPSF to ImageModels in romancal. 

Resolves [RCAL-573](https://jira.stsci.edu/browse/RCAL-573)

Closes #781

#### New dependencies

The model PSF is computed by `webbpsf`, and the model PSF is fit to Roman ImageModels using the new `photutils.psf` module refactor in https://github.com/astropy/photutils/pull/1558 (and futher PRs that followed it). I've pinned `photutils` to the dev version until the next release (expected within about a week), and pinned to the latest version of `webbpsf`.

Like CRDS, WebbPSF depends on several reference files which must be configured on a given machine before tests can be run. I've updated the CI setup so that GitHub Actions has access to those files. One limitation is that these WebbPSF reference files are versioned along with the software, so there are no dev files easily available to test against. For this reason, I've left WebbPSF out of the devdeps tests.

The last new dependency is `scikit-learn`. For crowded groups of sources, photutils has a `SourceGrouper` class, which uses a clustering algorithm implemented in `scikit-learn` to determine which sources are close enough to fit simultaneously with a joint model.

#### Implementation

There are two public methods introduced in this PR:

* `create_gridded_psf_model` uses WebbPSF to construct an ePSF model that is compatible with the photutils `psf` module. The function primarily serves to provide the user with reasonable defaults based on [previous testing](https://innerspace.stsci.edu/pages/viewpage.action?spaceKey=SCSB&title=Roman+astrometry+via+PSF+fitting+for+Source+Detection+Step). 

* `fit_psf_to_image_model` uses fits the PSF model to a Roman ImageModel. At this time, defaults have been optimized to give PSF fits with a good tradeoff between runtime and astrometric accuracy. Further tests should be done in the future to test photometric accuracy.

Tests are included to ensure that the inferred centroids of bright, high S/N, synthetic point sources are within 3-sigma of their injected values.

**Checklist**
- [ ] added entry in `CHANGES.rst` under the corresponding subsection
- [ ] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [ ] added relevant label(s)

### CODE REVIEW

**Comment by bmorris3** [comment]
@zacharyburnett @WilliamJamieson – it appears the failure on [py39-oldestdeps-cov](https://github.com/spacetelescope/romancal/actions/runs/5716259309/job/15487427936?pr=794) comes from the [photutils minpin for numpy](https://github.com/spacetelescope/romancal/actions/runs/5716259309/job/15487427936?pr=794#step:10:169) >= 1.22, while the others usually require >= 1.20. Any tips on what to do there?

**Comment by WilliamJamieson** [comment]
> @zacharyburnett @WilliamJamieson – it appears the failure on [py39-oldestdeps-cov](https://github.com/spacetelescope/romancal/actions/runs/5716259309/job/15487427936?pr=794) comes from the [photutils minpin for numpy](https://github.com/spacetelescope/romancal/actions/runs/5716259309/job/15487427936?pr=794#step:10:169) >= 1.22, while the others usually require >= 1.20. Any tips on what to do there?

I don't know why `romancal` has not updated its `numpy` lower pin. NumPy <= 1.21 are no longer supported according to NEP 29. @zacharyburnett should probably look into this.

**Comment by zacharyburnett** [comment]
pinned in https://github.com/spacetelescope/romancal/pull/800

File: pyproject.toml [hunk 1]
```diff
@@ -19,7 +19,8 @@ dependencies = [
     'gwcs >=0.18.1',
     'jsonschema >=4.0.1',
     'numpy >=1.22',
-    'photutils >=1.6.0',
+    # temporary pin to photutils dev until >1.8.0 is available:
+    'photutils @ git+https://github.com/astropy/photutils.git',
```
Comment (bmorris3) [inline pyproject.toml:23]: The next release of photutils may be out within a week or so, and I will update this pin when it's available.

File: romancal/lib/tests/test_psf.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        psf_model=psf_model,
+        x_init=true_x,
+        y_init=true_y,
+        fit_shape=fit_shape,
+    )
+
+    # difference between input and output, normalized by the
+    # uncertainty. Has units of sigma:
+    delta_x = np.abs(true_x - results_table["x_fit"]) / results_table["x_err"]
+    delta_y = np.abs(true_y - results_table["y_fit"]) / results_table["y_err"]
```
Comment (schlafly) [inline romancal/lib/tests/test_psf.py:154]: Can we also assert something about the size of x_err / y_err?  Maybe x_err < 1 / (amp / (sqrt(noise^2 + amp)) * fudge factor?  I just want to be able to catch failure modes like "all of the errors are very large, so we have small delta_x / delta_y but did a poor fit?  I'm trying to use the rule-of-thump x_err ~ FWHM / (S/N), which doesn't apply for this undersampled data but isn't terrible, and using 1 pix for the FWHM, and doing a bad approximation to the S/N.

**Review by schlafly — COMMENTED** [review]
This looks great to me.

Some things to think about:
- If we didn't use IterativePSFPhotometry, could we avoid the scikit-learn dependency?  It is pretty big and people might push back.  On the other hand, it would be nice to be able to use the IterativePSFPhotometry eventually.  I don't actually see it in the code, but I had thought that source grouping was a crowded field feature and we could get by with single star fits here.
- Re romancal vs. stcal vs. photutils: do you think fit_psf_to_image_model(...) would be sensible as part of photutils?
- dq_to_boolean_mask might fit better in romancal.lib.dqflags, but I don't feel strongly.
- There's a world where we precompute the gridded PSFs for each filter / SCA combination and save them somewhere (e.g., in CRDS).  How long is that computation?
- Does your wiki page use this code?  Can you link to it as part of this PR?
- Can you provide a little bit of run time information?
- I do think we need to understand the flipping of the stamps, though not as part of this PR.  But let's at least make an issue.
- This probably more naturally belongs under source_detection/psf.py rather than lib/psf.py?

File: romancal/lib/psf.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+filter_central_wavelengths = {
+    "WFI_Filter_F062_Center": 0.620,
+    "WFI_Filter_F087_Center": 0.869,
+    "WFI_Filter_F106_Center": 1.060,
+    "WFI_Filter_F129_Center": 1.293,
+    "WFI_Filter_F146_Center": 1.464,
+    "WFI_Filter_F158_Center": 1.577,
+    "WFI_Filter_F184_Center": 1.842,
+    "WFI_Filter_F213_Center": 2.125,
+}
```
[the remark this answers is not in the document]
Comment (nden) [reply romancal/lib/psf.py:42]: Is this info a good candidate for a reference file in CRDS? Then it can be changed by end users or INS without changing the code.

[the remark this answers is not in the document]
Comment (schlafly) [reply romancal/lib/psf.py:42]: It's not crazy.  I think if we wanted to go this route, we should instead get the whole filter curves into CRDS.  i.e. what Goddard has here:
https://roman.gsfc.nasa.gov/science/RRI/Roman_effarea_20201130.xlsx
If we had that kind of thing we could also put other summary information about the bandpasses in those files; e.g., central wavelengths.    The existing 'photom' file is some part of the way toward that end; one option would be to add more attributes there.  Obviously, let's not make that part of this PR, though!

[the remark this answers is not in the document]
Comment (bmorris3) [reply romancal/lib/psf.py:42]: Great, some options for moving forward: shall we make a ticket to add these quantities to CRDS, reference that ticket in a `TODO` comment in the file in this PR, and make an RCAL ticket to do a follow-up PR when those quantities are available?

File: romancal/lib/psf.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+            f"Loading existing gridded PSF model from {expected_output_path}",
+        )
+        psf_model = CCDData.read(expected_output_path, unit=u.electron / u.s, ext=0)
+        # the FITS file saved by webbpsf gets loaded with pixel
+        # axes flipped in both dimensions, so flip them back after loading:
+        psf_model.data = psf_model.data[::-1, ::-1]
+        psf_model.meta = dict(psf_model.meta)
+        psf_model.meta["oversampling"] = oversample
+        psf_model.meta["grid_xypos"] = np.array(
+            [list(tup)[::-1] for tup in model_psf_centroids]
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply romancal/lib/psf.py:179]: This seems to come from webbpsf storing `DET_YX` coordinates, and photutils expecting `grid_xypos` coordinates in the `NDData` object (see https://github.com/spacetelescope/webbpsf/blob/34ba8fc08a71148447a7ccf970b32abcd93f60a2/webbpsf/gridded_library.py#L437-L439C66).

File: romancal/lib/psf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nden) [reply romancal/lib/psf.py:42]: sounds good to me

**Comment by bmorris3** [comment]
RE @schlafly:

> If we didn't use IterativePSFPhotometry, could we avoid the scikit-learn dependency?  It is pretty big and people might push back.  On the other hand, it would be nice to be able to use the IterativePSFPhotometry eventually.  I don't actually see it in the code, but I had thought that source grouping was a crowded field feature and we could get by with single star fits here.

Yes! The additional `scikit-learn` is avoidable if we don't use IterativePSFPhotometry as it's implemented in photutils right now. But it's also possible(/likely?) that other methods like `photutils.psf.SourceGrouper`, or clustering algorithms within that method, will be implemented that don't depend on the `scikit-learn` (cc @larrybradley). There's a growing number of options available in [scipy.cluster](https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html#module-scipy.cluster.hierarchy) that we can experiment with in photutils to dodge this big dependency without loosing the iterative fitting functionality. 

> Re romancal vs. stcal vs. photutils: do you think fit_psf_to_image_model(...) would be sensible as part of photutils?

The two methods in this PR are pretty simple – they allow a user fit PSF models to Roman ImageModels. Almost all of the work in this PR was choosing and vetting sensible defaults for the input parameters for PSF fitting specifically for Roman ImageModels. So personally, I think the Roman-specificity of this PR is the primary feature of this PR, making the other repos ill-fitting homes for this PR.

> dq_to_boolean_mask might fit better in romancal.lib.dqflags, but I don't feel strongly.

I agree, that's fair. The only reason I've put it in `lib.psf` here is because the Roman dq flag map defines `"GOOD": 0`, which needs to be removed before extending the astropy bitmask registry, in order to work correctly for photutils. I suppose that's true for anyone who expects a boolean mask though.

> There's a world where we precompute the gridded PSFs for each filter / SCA combination and save them somewhere (e.g., in CRDS).  How long is that computation?

This depends on a few parameters, primarily: (1) the number of PSFs to synthesize across each SCA, (2) the width of the FOV of the rendered PSF. The default gridded PSF model from this PR has a FOV of 50 pixels, rendered in 16 locations distributed across the SCA, and takes 10 seconds to run. For all SCAs, runtime will be a few minutes.

> Does your wiki page use this code?  Can you link to it as part of this PR?

I wrote up a performance evaluation for the code in [this repo](https://github.com/bmorris3/schmear) in the [confluence post](https://ssoportal.stsci.edu/idp/profile/SAML2/POST/SSO). The methods in this PR are essentially identical with two exceptions. This PR provides sensible defaults given the results of the analysis, and this PR has comments/docstrings/tests.    

> Can you provide a little bit of run time information?

With the default settings in this PR, I've run both methods for an exposure the F087 filter on SCA 7. Constructing a gridded PSF model takes 10 seconds. I ran the PSF model fit on a synthetic ImageModel, without supplying initial centroid or flux guesses. Of the 500 sources injected with `romanisim`, the default finder (`DAOStarFind`) detected 250 sources and fit their PSFs in 2.7 seconds.

> I do think we need to understand the flipping of the stamps, though not as part of this PR.  But let's at least make an issue.

Agreed.

> This probably more naturally belongs under source_detection/psf.py rather than lib/psf.py?

I'm happy either way. The forward-looking intention for putting it in `romancal.lib` is to separate it from Source Detection in anticipation of a future photometry step that also uses PSF photometry.

**Comment by bmorris3** [comment]
The devdeps failure is related to a round-trip parsing issue in WebbPSF, I've posted an issue there: https://github.com/spacetelescope/webbpsf/issues/710

**Comment by larrybradley** [comment]
@bmorris3 I've updated `photutils.psf.SourceGrouper` to use `scipy.cluster.hierarchy.fclusterdata` instead of `sklearn.cluster.DBSCAN`.  In my testing, they give the same results.  `scikit-learn` is no longer needed for source grouping.

**Comment by bmorris3** [comment]
In the time since I posted this PR, I found [the GSFC document](https://roman.ipac.caltech.edu/sims/Param_db.html) that outlines where this jitter number comes from (search for `Telescope_Jitter`). Since this is a GSFC benediction, I'll opt to close this PR, though maybe in the future we should document that this assumption is made upstream in WebbPSF.

**Comment by schlafly** [comment]
Did you mean to close https://github.com/spacetelescope/romanisim/pull/63 instead of this one?

**Comment by bmorris3** [comment]
> Did you mean to close [spacetelescope/romanisim#63](https://github.com/spacetelescope/romanisim/pull/63) instead of this one?

Yikes you're right! Thanks @schlafly!

File: pyproject.toml [hunk 5]
```diff
@@ -19,7 +19,7 @@ dependencies = [
     'gwcs >=0.18.1',
     'jsonschema >=4.8',
     'numpy >=1.22',
-    'photutils >=1.6.0',
+    'photutils @ git+https://github.com/astropy/photutils.git',
```
Comment (schlafly) [inline pyproject.toml:22]: It looks to me like this is likely out now?

File: .github/workflows/roman_ci.yml [hunk 6]
```diff
@@ -39,10 +46,27 @@ jobs:
...
-      context: ${{ steps.context.outputs.pmap }}
-      path: ${{ steps.path.outputs.path }}
-      server: ${{ steps.server.outputs.url }}
+      # webbpsf:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - id: data
+        run: |
+          echo "webbpsf_url=https://stsci.box.com/shared/static/n1fealx9q0m6sdnass6wnyfikvxtc0zz.gz" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (schlafly) [reply .github/workflows/roman_ci.yml:56]: No changes needed here, but for romanisim it's annoying that we have to update this each time a new webbpsf release happens.  I haven't understood why this file can't have a name like webbpsf-latest-release-test.gz or something, and when a new version comes out the new test data file gets updated.  That would mean that each new version doesn't require new CI updates.

**Review by schlafly — COMMENTED** [review]
Thanks.  Two more minor comments.  I don't think we have any regression tests here, so I can't see how this would have an effect there, but we usually require runs of the regression suite with the changes before merging.

File: pyproject.toml (hunk 5 above)
Comment (bmorris3) [reply pyproject.toml:22]: The last remaining and necessary addition to photutils is from https://github.com/astropy/photutils/pull/1617. I'll ask @larrybradley if the next release is likely to be soon.

File: romancal/lib/tests/test_psf.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    synth_err = image_model.err
+    psf_shape = np.array(psf_model.data.shape[1:]) // oversample
+
+    for fit_model in fit_models:
+        x0 = getattr(fit_model, xname).value
+        y0 = getattr(fit_model, yname).value
+        slc_lg, _ = overlap_slices(synth_image.shape, psf_shape, (y0, x0), mode="trim")
+        yy, xx = np.mgrid[slc_lg]
+        model_data = fit_model(xx, yy) * image_model.data.unit
+        model_err = np.sqrt(model_data.value) * model_data.unit
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply romancal/lib/tests/test_psf.py:91]: Good point. So far I'm not yet adding the shot noise to the data array, but I am increasing the uncertainties as the counts change.

Write
Preview
Attach files by dragging & dropping, selecting or pasting them.
Cancel
Update comment

**Review by nden — APPROVED** [review]
